crates/ty_python_semantic/resources/mdtest/generics/specialize_constrained.md

```diff
@@ -0,0 +1,273 @@
+    reveal_type(generic_context(bounded_by_gradual).specialize_constrained(ConstraintSet.range(Never, T, Base)))
+
+    # revealed: ty_extensions.Specialization[T@bounded_by_gradual = Unrelated]
+    reveal_type(generic_context(bounded_by_gradual).specialize_constrained(ConstraintSet.range(Never, T, Unrelated)))
+
+def bounded_by_gradual_list[T: list[Any]]():
+    # revealed: ty_extensions.Specialization[T@bounded_by_gradual_list = Top[list[Any]]]
+    reveal_type(generic_context(bounded_by_gradual_list).specialize_constrained(ConstraintSet.always()))
+    # revealed: None
+    reveal_type(generic_context(bounded_by_gradual_list).specialize_constrained(ConstraintSet.never()))
+
+    # revealed: ty_extensions.Specialization[T@bounded_by_gradual_list = list[Base]]
+    reveal_type(generic_context(bounded_by_gradual_list).specialize_constrained(ConstraintSet.range(Never, T, list[Base])))
+
+    # revealed: ty_extensions.Specialization[T@bounded_by_gradual_list = list[Unrelated]]
+    reveal_type(generic_context(bounded_by_gradual_list).specialize_constrained(ConstraintSet.range(Never, T, list[Unrelated])))
+```
+
+## Constrained typevar
+
+If a typevar has constraints, then it must specialize to one of those specific types. (Not to a
+subtype of one of those types!)
+
+```py
+from typing import final, Never
+from ty_extensions import ConstraintSet, generic_context
+
+class Super: ...
+class Base(Super): ...
+class Sub(Base): ...
+
+@final
+class Unrelated: ...
+
+def constrained[T: (Base, Unrelated)]():
+    # revealed: ty_extensions.Specialization[T@constrained = Base | Unrelated]
+    reveal_type(generic_context(constrained).specialize_constrained(ConstraintSet.always()))
+    # revealed: None
+    reveal_type(generic_context(constrained).specialize_constrained(ConstraintSet.never()))
+
+    # revealed: ty_extensions.Specialization[T@constrained = Base]
+    reveal_type(generic_context(constrained).specialize_constrained(ConstraintSet.range(Never, T, Base)))
+    # revealed: ty_extensions.Specialization[T@constrained = Unrelated]
+    reveal_type(generic_context(constrained).specialize_constrained(ConstraintSet.range(Never, T, Unrelated)))
+
+    # revealed: ty_extensions.Specialization[T@constrained = Base]
+    reveal_type(generic_context(constrained).specialize_constrained(ConstraintSet.range(Never, T, Super)))
+    # revealed: None
+    reveal_type(generic_context(constrained).specialize_constrained(ConstraintSet.range(Super, T, Super)))
+
+    # revealed: ty_extensions.Specialization[T@constrained = Base]
+    reveal_type(generic_context(constrained).specialize_constrained(ConstraintSet.range(Sub, T, object)))
+    # revealed: None
+    reveal_type(generic_context(constrained).specialize_constrained(ConstraintSet.range(Sub, T, Sub)))
+```
+
+If any of the constraints is a gradual type, we are free to choose any materialization of that
+constraint that makes the test succeed.
```

Comment:
Yes, but shouldn't we still solve to the gradual type that is one of the constraints? E.g. see https://pyright-play.net/?pyrightVersion=1.1.405&pythonVersion=3.12&reportUnreachable=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcAUCQCYCmwUwA2gCoBcUAFKjADSHECUAuqwAeLBlF5QAtAD4ozElEVQQlGAFcQKKELIqAbpQCGAGwD68BJVbBWAImBgwt3rzJA

E.g. in every solution in the first section below that is not `Base`, I would expect to solve to `Any`, rather than any of the materializations of it. I suspect this will cause false positives and violate the gradual guarantee otherwise?

It seems like the current approach takes as a requirement that we always solve to a fully static type, but I'm not sure that is what we should do, in the case of constrained type variables.